topics/reference_guide/custom_language_support/websymbols_context.md

```diff
@@ -0,0 +1,257 @@
+from the one acquired on a `PsiElement` location. This is expected. The `VirtualFile` location is used to,
+amongst others, determine a language to parse the file, so it cannot use a PSI tree. It must also be rapid,
+since it's used during the indexing phase.
+
+Usually, the coding assistance logic is working with `PsiElement`, so you should provide it as the location
+at which context is checked. It is important, though, to remember that the language of the file might only
+be substituted if a context is detected on `VirtualFile` level.
+
+A good example of how context detection can be used is web frameworks, since at a particular location only
+one of the frameworks can be used. Various `WebSymbolsContextProvider` extensions and context rules are used
+to determine which of the frameworks (Vue, Angular, React, Astro, etc.) to enable at the particular location.
+And each of the plugins calls `WebSymbolsContext.get("framework", ...)` method to check if it's their framework,
+which is enabled.
+
+## `WebSymbolsContextProvider`
+
+The most straightforward way to contribute context is to register a `WebSymbolsContextProvider`
+through `com.intellij.webSymbols.context` and override one of `isEnabled` methods
+to detect a context of a particular name and kind, e.g.:
+
+```xml
+<idea-plugin>
+  <extensions defaultExtensionNs="com.intellij">
+    <webSymbols.context kind="stimulus-context" name="true" implementation="com.intellij.stimulus.context.StimulusContextProvider"/>
+  </extensions>
+</idea-plugin>
+```
+
+The `StimulusContextProvider` can, for instance, check for JavaScript imports in the file to see if some additional Stimulus support
+should be present in the file. The calculations should be cached, as `WebSymbolsContextProvider` methods are called very often.
+
+The presence of the context can be checked as follows:
+
+```kotlin
+WebSymbols.get("stimulus-context", psiElement) == "true"
+```
+
+`WebSymbolsContextProvider` can also prevent a context from being detected. In this case, you should override `isForbidden` method.
+To prevent any context of a particular kind, use `any` as a name. E.g.:
+
+```xml
+<idea-plugin>
+  <extensions defaultExtensionNs="com.intellij">
+    <webSymbols.context kind="framework" name="any" implementation="com.intellij.python.js.PyTemplatesWebContextBlocker"/>
+  </extensions>
+</idea-plugin>
+```
+
+In this example, the `PyTemplatesWebContextBlocker` can check if the project has any Python templating engine enabled in the
+location and forbid the `framework` context to disable web frameworks support, like Angular or Vue, which conflict with e.g., Blade support.
+
+## Context rules
+
+`WebSymbolsContextProvider` is straightforward to use, but it is not very efficient. When many providers look for similar information,
+a lot of calculations are repeated and the performance of the whole IDE drops down. One of the examples is when providers look for the presence
```

Comment:
a lot of calculations are repeated affecting the overall performance of the IDE.